File: Insert_Delete_Getrandom_O(1).java
/*380. Insert Delete GetRandom O(1)  QuestionEditorial Solution  My Submissions
Total Accepted: 4173
Total Submissions: 11983
Difficulty: Hard
Design a data structure that supports all following operations in average O(1) time.

insert(val): Inserts an item val to the set if not already present.
remove(val): Removes an item val from the set if present.
getRandom: Returns a random element from current set of elements. Each element must have the same probability of being returned.
Example:

// Init an empty set.
RandomizedSet randomSet = new RandomizedSet();

// Inserts 1 to the set. Returns true as 1 was inserted successfully.
randomSet.insert(1);

// Returns false as 2 does not exist in the set.
randomSet.remove(2);

// Inserts 2 to the set, returns true. Set now contains [1,2].
randomSet.insert(2);

// getRandom should return either 1 or 2 randomly.
randomSet.getRandom();

// Removes 1 from the set, returns true. Set now contains [2].
randomSet.remove(1);

// 2 was already in the set, so return false.
randomSet.insert(2);

// Since 1 is the only number in the set, getRandom always return 1.
randomSet.getRandom();
Hide Company Tags Google Uber Twitter Amazon Yelp Pocket Gems
Hide Tags Array Hash Table Design
Hide Similar Problems (H) Insert Delete GetRandom O(1) - Duplicates allowed
*/




/*
use HashMap<Integer,Integer> mymap, key is cur val, value is when it insert, which is ele.size(), for convenient to check if the key is exist in map
ArrayList<Integer> ele store the current key we already inserted
insert(val) --- mymap.containsKey(val), if it is, return false, else insert it
remove(val) --- first check if this val exist, if it isnot exist, return false
    then, remove if from mymap and ele list
    before remove it ,
        to ele list, switch cur val to the end of ele list
        to mymap, also switch the its value with the end value
getRandom --- use java.util.Random to random get the random ele in ele list
    ele.get(rand.nextInt(ele.size()))*/
class RandomizedSet {
    List<Integer> arr;
    HashMap<Integer, Integer> mymap;
    java.util.Random rand = new java.util.Random();

    public RandomizedSet() {
        arr = new ArrayList<>();
        mymap = new HashMap<>();
    }

    public boolean insert(int val) {
        if (mymap.containsKey(val)) {
            return false;
        } else {
            mymap.put(val, arr.size());
            arr.add(val);
            return true;
        }
    }

    public boolean remove(int val) {
        if (!mymap.containsKey(val)) {
            return false;
        } else {
            int index = mymap.get(val);
            if (index < arr.size()-1) {
                int lastElement = arr.get(arr.size() - 1);
                arr.set(index, lastElement);
                mymap.put(lastElement, index);
            }
            arr.remove(arr.size()-1);
            mymap.remove(val);
            return true;
        }
    }

    public int getRandom() {
        return arr.get(rand.nextInt(arr.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: Insert_Delete_Getrandom_O(1).py
'''
Implement the RandomizedSet class:

RandomizedSet() Initializes the RandomizedSet object.
bool insert(int val) Inserts an item val into the set if not present. Returns true if the item was not present, false otherwise.
bool remove(int val) Removes an item val from the set if present. Returns true if the item was present, false otherwise.
int getRandom() Returns a random element from the current set of elements (it's guaranteed that at least one element exists when this method is called). Each element must have the same probability of being returned.
You must implement the functions of the class such that each function works in average O(1) time complexity.



Example 1:

Input
["RandomizedSet", "insert", "remove", "insert", "getRandom", "remove", "insert", "getRandom"]
[[], [1], [2], [2], [], [1], [2], []]
Output
[null, true, false, true, 2, true, false, 2]

Explanation
RandomizedSet randomizedSet = new RandomizedSet();
randomizedSet.insert(1); Inserts 1 to the set. Returns true as 1 was inserted successfully.
randomizedSet.remove(2); Returns false as 2 does not exist in the set.
randomizedSet.insert(2); Inserts 2 to the set, returns true. Set now contains [1,2].
randomizedSet.getRandom(); getRandom() should return either 1 or 2 randomly.
randomizedSet.remove(1); Removes 1 from the set, returns true. Set now contains [2].
randomizedSet.insert(2); 2 was already in the set, so return false.
randomizedSet.getRandom(); Since 2 is the only number in the set, getRandom() will always return 2.


Constraints:

-231 <= val <= 231 - 1
At most 2 * 105 calls will be made to insert, remove, and getRandom.
There will be at least one element in the data structure when getRandom is called.
'''

'''
Solution 2: hashmap & arr

Insert
Add value -> its index into dictionary, average O(1) time.
Append value to array list, average O(1) time as well.

Delete
Retrieve an index of element to delete from the hashmap.
Move the last element to the place of the element to delete, O(1) time.
Pop the last element out, O(1) time.

GetRandom
GetRandom could be implemented in O(1) time with the help of standard random.choice in Python and Random object in Java.

Time complexity. GetRandom is always O(1). Insert and Delete both have O(1) average time complexity, and \mathcal{O}(N)O(N) in the worst-case scenario when the operation exceeds the capacity of currently allocated array/hashmap and invokes space reallocation.
Space complexity: O(N), to store N elements.
'''
class RandomizedSet:

    def __init__(self):
        self.mymap = {}
        self.arr = []

    def insert(self, val: int) -> bool:
        if val in self.mymap: return False
        self.arr.append(val)
        self.mymap[val] = len(self.arr)-1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.mymap: return False
        # put last element into val element position
        lastEle = self.arr[-1]
        valIndex = self.mymap[val]
        self.mymap[lastEle] = valIndex
        self.arr[valIndex] = lastEle
        # remove last element
        self.arr.pop(-1)
        self.mymap.pop(val)
        return True

    def getRandom(self) -> int:
        return self.arr[randint(0, len(self.arr)-1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
